[Core] Support NIXL storage obj backend

tshmilnvidia opened the pull request:
Support NIXL obj backend
    
    Adding support for NIXL obj backend.
    
    in order to use this backend, AWS parameters should be provided via
    extra_config.nixl_backend_params in the yaml file, for example:
    
    extra_config:
      enable_nixl_storage: true
      nixl_backend: OBJ
      nixl_pool_size: 64
      nixl_path: /mnt/nixl/cache/
      nixl_backend_params:
        access_key: <your_access_key>
        secret_key: <your_secret_key>
        bucket: <your_bucket>
        region: <your_region>

tshmilnvidia committed:
[Core] Generalize NixlFilePool

Signed-off-by: Tomer Shmilovich <tshmilovich@nvidia.com>

tshmilnvidia commented on lmcache/v1/storage_backend/nixl_storage_backend.py:
```diff
@@ -126,38 +171,57 @@ def close(self):
+            key = f"obj_{i}_{uuid.uuid4().hex[0:4]}.bin"
+            self.keys.append(key)
+
+    def close(self):
+        pass
```
nothing to clear here

Shaoting-Feng reviewed (approved):
LGTM

tshmilnvidia commented:
Hi @Shaoting-Feng, please notice I added a few minor changes, mostly to address comments left by the AI assistant.

Shaoting-Feng commented:
Please resolve the conflicts.

tshmilnvidia commented:
> Please resolve the conflicts.

Done

Shaoting-Feng reviewed (approved):
LGTM

tshmilnvidia committed:
[Core] NixlStorageBackend support backend params

Signed-off-by: Tomer Shmilovich <tshmilovich@nvidia.com>

tshmilnvidia committed:
[Core] Support NixlStorageBackend OBJ backend

Adding support for NIXL obj backend.

in order to use this backend, AWS parameters should be provided via
extra_config.nixl_backend_params in the yaml file, for example:

extra_config:
  enable_nixl_storage: true
  nixl_backend: OBJ
  nixl_pool_size: 64
  nixl_path: /mnt/nixl/cache/
  nixl_backend_params:
    access_key: <your_access_key>
    secret_key: <your_secret_key>
    bucket: <your_bucket>
    region: <your_region>

Signed-off-by: Tomer Shmilovich <tshmilovich@nvidia.com>

tshmilnvidia committed:
[Docs] Update docs

Signed-off-by: Tomer Shmilovich <tshmilovich@nvidia.com>

tshmilnvidia commented:
Hi @Shaoting-Feng, we had some internal talks and decided to change our approach to how we provide the backend specific params. Modified this PR in accordance and also updated docs.

ApostaC commented:
@tshmilnvidia I remember in the previous code there are AWS key configurations. Are those configurations not needed anymore? What should users config if they want to use AWS s3?

tshmilnvidia commented:
> @tshmilnvidia I remember in the previous code there are AWS key configurations. Are those configurations not needed anymore? What should users config if they want to use AWS s3?

They are still needed, I have modified the first comment of this PR to serve as an example. I have also added this to the commit message of the relevant commit.

sammshen reviewed (approved):
LGTM!

sammshen reviewed (changes requested):
please run the precommit for code quality

tshmilnvidia commented:
> please run the precommit for code quality

I believe these errors are unrelated and were fixed here:
https://github.com/LMCache/LMCache/pull/1634

I'll rebase

ApostaC commented on docs/source/kv_cache/storage_backends/nixl.rst:
```diff
@@ -35,16 +35,22 @@ Example ``lmcache-config.yaml``:
+
+    Supported backends are: ["GDS", "GDS_MT", "POSIX", "HF3FS", "OBJ"].
+
+    Backend specific params should be provided via ``extra_config.nixl_backend_params``. Please refer to NIXL documentation for specifics.
```
Can we be more specific here? What are the params needed? 

Or at least please give a link to the specific NIXL documentation. I'm not able to quickly find it online.

tshmilnvidia replied:
https://github.com/ai-dynamo/nixl/tree/main/src/plugins/obj

Link for specific example of NIXL OBJ backend documentation.

Regarding naming - OBJ is how this backend is called in NIXL, and it is intended to support generic object storage, not only S3 - we want to stick with NIXL naming here.

I will update the doc with a specific S3 example.

ApostaC reviewed (commented):
LGTM in general. Two high-level comments:
- Is the `OBJ` backend designed for S3 API? If so, should we name it as "S3" for better clarity?
- Can we update the doc to include the instructions you had in the PR description, so that the users will know how to use it with S3.

ziruiliu commented:
Hi @tshmilnvidia, I have a question for both files and objects in nixl_storage_backend: is there any suggested value of nixl_pool_size? should it be aligned with local cache size? 
To my understanding, when cache data increases, file_pool will be used up and become empty. calling remove() will put files back to file_pool. Is there any utilily that manually triggers remove()? I did not find any automatic remove/eviction in storage manager.

tshmilnvidia commented:
> Hi @tshmilnvidia, I have a question for both files and objects in nixl_storage_backend: is there any suggested value of nixl_pool_size? should it be aligned with local cache size? To my understanding, when cache data increases, file_pool will be used up and become empty. calling remove() will put files back to file_pool. Is there any utilily that manually triggers remove()? I did not find any automatic remove/eviction in storage manager.

@ziruiliu 
I don't have a specific value in mind, I think it's best to experiment and see what works best for each case.

Your understanding regarding files being used up is correct, but I believe storage manager does call batched_remove(), which in turn calls remove() via the abstract_backend implementation of batched_remove(). In any case, I'm working on integrating cache policy eviction which should help with this as well.

jgoldsch12 commented on lmcache/v1/storage_backend/nixl_storage_backend.py:
```diff
@@ -357,7 +437,7 @@ async def file_to_gpu(
             return obj_list
 
-        handle = self.agent.get_file_to_gpu_handle(mem_indices, file_indices)
+        handle = self.agent.get_storage_to_mem_handle(mem_indices, file_indices)
```
file_indices could also use a generic name

tshmilnvidia replied:
Done

jgoldsch12 commented on tests/v1/data/nixl.yaml:
```diff
@@ -3,4 +3,4 @@ chunk_size: 256
 nixl_buffer_size: 1073741824
 nixl_buffer_device: cpu
-extra_config: {enable_nixl_storage: true, nixl_backend: POSIX, nixl_file_pool_size: 64, nixl_path: /tmp/nixl/cache/}
+extra_config: {enable_nixl_storage: true, nixl_backend: POSIX, nixl_pool_size: 64, nixl_path: /tmp/nixl/cache/}
```
I'd recommend the same YAML reformatting added to the docs applied here.

tshmilnvidia replied:
Done

tshmilnvidia commented:
Rebased

sammshen commented:
LGTM, can you add the nixl.rst to teh docs index so it will actually appear, thanks!

tshmilnvidia commented:
> LGTM, can you add the nixl.rst to teh docs index so it will actually appear, thanks!

@sammshen 

Are you referring to this?

https://github.com/LMCache/LMCache/blob/011edff4bd7191f63f519598f5e108f12f610ae5/docs/source/kv_cache/storage_backends/index.rst

it's already there, this docs file is not new, it is only updated.

sammshen commented:
thanks @tshmilnvidia

mmgaggle commented:
Thank you for this, we were working around this with https://github.com/mmgaggle/LMCache/tree/wip-nixl-obj
